src/lib/modules/specialconfigs/index.js

```diff
@@ -133,54 +143,128 @@ class SpecialConfigs {
-          return cb(new Error("error running onDeploy for " + contract.className.cyan));
-        }
 
-        self.runOnDeployCode(onDeployCode, cb, contract.silent);
-      });
+          self.runOnDeployCode(onDeployCode, cb, contract.silent);
+        });
+      }
     });
   }
 
   registerDeployIfAction() {
     const self = this;
 
-    self.embark.registerActionForEvent("deploy:contract:shouldDeploy", (params, cb) => {
+    self.embark.registerActionForEvent("deploy:contract:shouldDeploy", async (params, cb) => {
       let cmd = params.contract.deployIf;
+      const contract = params.contract;
       if (!cmd) {
         return cb(params);
       }
 
-      self.events.request('runcode:eval', cmd, (err, result) => {
+      if (typeof cmd === 'function') {
+        try {
+          const dependencies = await this.getOnDeployLifecycleHookDependencies(contract);
+          params.shouldDeploy = await contract.deployIf(dependencies);
+          cb(params);
+        } catch (err) {
+          return cb(new Error(`Error when registering deployIf hook for ${contract.name}: ${err.message}`));
+        }
+      } else {
+
+        self.events.request('runcode:eval', cmd, (err, result) => {
+          if (err) {
+            self.logger.error(params.contract.className + ' deployIf directive has an error; contract will not deploy');
+            self.logger.error(err.message || err);
+            params.shouldDeploy = false;
+          } else if (!result) {
+            self.logger.info(params.contract.className + ' deployIf directive returned false; contract will not deploy');
+            params.shouldDeploy = false;
+          }
+
+          cb(params);
+        });
+      }
+    });
+  }
+
+  getOnDeployLifecycleHookDependencies(contractConfig) {
+    let dependencyNames = contractConfig.deps || [];
+    dependencyNames.push(contractConfig.className);
+
+    return new Promise((resolve, reject) => {
+      async.map(dependencyNames, (contractName, next) => {
+        this.events.request('contracts:contract', contractName, (contractRecipe) => {
+          if (!contractRecipe) {
+            next(new Error('ReferredContractDoesNotExist'));
+          }
+          this.events.request('blockchain:contract:create', {
+            abi: contractRecipe.abiDefinition,
+            address: contractRecipe.deployedAddress
+          }, contractInstance => {
+            next(null, { className: contractRecipe.className, instance: contractInstance });
+          });
+        });
+      }, (err, contractInstances) => {
         if (err) {
-          self.logger.error(params.contract.className + ' deployIf directive has an error; contract will not deploy');
-          self.logger.error(err.message || err);
-          params.shouldDeploy = false;
-        } else if (!result) {
-          self.logger.info(params.contract.className + ' deployIf directive returned false; contract will not deploy');
-          params.shouldDeploy = false;
+          reject(err);
```

Comment:
I think you don't have to return here, but I feel like it's safer to anyway.

Reply:
Correct, no return statement needed for `resolve/reject` as this will return the promise immediately.